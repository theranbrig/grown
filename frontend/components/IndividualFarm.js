import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import Head from 'next/head';
import dynamic from 'next/dynamic';
import styled from 'styled-components';
import { Loader, Grid, Image, Icon } from 'semantic-ui-react';
import MapContainer from './FarmMap';
import { GoogleApiWrapper } from 'google-maps-react';
import StyledFarmInfo from './styles/IndividualFarmStyles';
import Link from 'next/link';
import User from './User';
import DeleteFarmButton from './DeleteFarmButton';
import Store, { PRODUCTS_QUERY } from './Store';
import CreateProduct from './CreateProduct';

const INDIVIDUAL_FARM_QUERY = gql`
	query INDIVIDUAL_FARM_QUERY($id: ID!) {
		farm(where: { id: $id }) {
			id
			name
			tagline
			description
			location
			phone
			email
			image
			user {
				id
			}
		}
	}
`;

class IndividualFarm extends Component {
	render() {
		return (
			<User>
				{({ data: { me } }) => (
					<Query query={INDIVIDUAL_FARM_QUERY} variables={{ id: this.props.id }}>
						{({ data, loading, error }) => {
							if (loading) return <Loader active inline />;
							if (error) return <p>Error...</p>;
							const farm = data.farm;
							return (
								<StyledFarmInfo>
									<div className="farm-header">
										<h2>{farm.name}</h2>
										<h3>Explore and Shop</h3>
									</div>
									<Grid container>
										<Grid.Row>
											<Grid.Column mobile={16} computer={8}>
												<Image src={farm.image} alt={farm.name} centered />
												<h3>{farm.tagline}</h3>
												<Query query={PRODUCTS_QUERY} variables={{ farmId: farm.id }}>
													{({ data, error, loading }) => {
														return (
															<ul>
																{data.products.slice(0, 5).map(product => (
																	<li key={product.id}>{product.name}</li>
																))}
															</ul>
														);
													}}
												</Query>
											</Grid.Column>
											<Grid.Column mobile={16} computer={8}>
												<div className="info-box">
													<div className="text-info">
														<h4>{farm.name}</h4>

														<p>
															<Icon name="map" />
															{farm.location}
														</p>
														<p>{farm.description}</p>
														<div className="phone-email">
															<Icon name="mail" />
															<a href={`mailto:${farm.email}`}>{farm.email}</a>
														</div>
														<div className="phone-email">
															<Icon name="phone" />
															<a href={`tel:${farm.phone}`}>{farm.phone}</a>
														</div>
													</div>
													<MapContainer location={farm.location} name={farm.name} />
												</div>
												{me &&
													me.id === farm.user.id && (
														<div className="edit-farm-link">
															<Link
																href={{
																	pathname: '/updatefarm',
																	query: { id: farm.id }
																}}>
																<a>
																	<Icon name="edit" />
																	Edit Farm Information
																</a>
															</Link>
															<DeleteFarmButton id={farm.id} />
														</div>
													)}
											</Grid.Column>
										</Grid.Row>
									</Grid>
									<Grid container centered>
										<Grid.Column width={16}>
											<Store id={farm.id} name={farm.name} />
											{me && me.id === farm.user.id && <CreateProduct id={farm.id} />}
										</Grid.Column>
									</Grid>
								</StyledFarmInfo>
							);
						}}
					</Query>
				)}
			</User>
		);
	}
}

export default IndividualFarm;
export { INDIVIDUAL_FARM_QUERY };
