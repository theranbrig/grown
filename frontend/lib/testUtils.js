import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

const fakeFarm = () => ({
	__typename: 'Farm',
	id: '1234567890',
	name: 'Farmy McFarmson',
	description: 'Great Stuff',
	location: 'The World',
	email: 'email@email.com',
	phone: '123-456-7890',
	image: 'farm.jpg',
	website: 'www.farm.com'
});

const fakeUser = () => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: []
});

const fakeOrderItem = () => ({
	__typename: 'OrderItem',
	id: casual.uuid,
	image: `${casual.word}.jpg`,
	title: casual.words(),
	price: 4234,
	quantity: 1,
	description: casual.words()
});

const fakeOrder = () => ({
	__typename: 'Order',
	id: 'ord123',
	charge: 'ch_123',
	total: 40000,
	items: [fakeOrderItem(), fakeOrderItem()],
	createdAt: '2018-04 - 06T19: 24: 16.000Z',
	user: fakeUser()
});

const fakeCartItem = overrides => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 3,
	item: fakeItem(),
	user: fakeUser(),
	...overrides
});

// Fake LocalStorage
class LocalStorageMock {
	constructor() {
		this.store = {};
	}

	clear() {
		this.store = {};
	}

	getFarm(key) {
		return this.store[key] || null;
	}

	setItem(key, value) {
		this.store[key] = value.toString();
	}

	removeItem(key) {
		delete this.store[key];
	}
}

export { LocalStorageMock, fakeFarm, fakeUser, fakeCartItem, fakeOrder, fakeOrderItem };
