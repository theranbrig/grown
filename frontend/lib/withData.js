import ApolloClient from 'apollo-boost';
import withApollo from 'next-with-apollo';
import { endpoint, prodEndpoint } from '../config';

function createClient({ headers }) {
	// Set up apollo client
	return new ApolloClient({
		uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
		request: operation => {
			operation.setContext({
				fetchOptions: {
					credentials: 'include'
				},
				headers
			});
		},
		// Local Data
		clientState: {
			resolvers: {}
		}
	});
}

export default withApollo(createClient);
