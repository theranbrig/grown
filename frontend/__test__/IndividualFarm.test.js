import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import IndividualFarm, { INDIVIDUAL_FARM_QUERY } from '../components/IndividualFarm';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeFarm } from '../lib/testUtils';

describe('Individual Farm Page', () => {
	it('should render the individual farm page', async () => {
		const mocks = [
			{
				request: { query: INDIVIDUAL_FARM_QUERY, variables: { id: '1234567890' } },
				result: {
					data: {
						farm: fakeFarm()
					}
				}
			}
		];
		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<IndividualFarm id="1234567890" />
			</MockedProvider>
		);
		await wait();
		wrapper.update();
		expect(toJSON(wrapper.find('h4'))).toMatchSnapshot();
		expect(toJSON(wrapper.find('img'))).toMatchSnapshot();
		expect(toJSON(wrapper.find('h2'))).toMatchSnapshot();
	});

	it('Errors with a not found farm', async () => {
		const mocks = [
			{
				request: { query: INDIVIDUAL_FARM_QUERY, variables: { id: '1234567890' } },
				result: {
					errors: [{ message: 'Farm Not Found' }]
				}
			}
		];
		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<IndividualFarm id="1234567890" />
			</MockedProvider>
		);
		await wait();
		wrapper.update();
		console.log(wrapper.debug());
		expect(toJSON(wrapper.find('p'))).toMatchSnapshot();
	});
});
