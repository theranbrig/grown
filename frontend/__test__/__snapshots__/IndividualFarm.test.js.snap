// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Individual Farm Page Errors with a not found farm 1`] = `
<p>
  Error...
</p>
`;

exports[`Individual Farm Page should render the individual farm page 1`] = `
<h4>
  Farmy McFarmson
</h4>
`;

exports[`Individual Farm Page should render the individual farm page 2`] = `
<img
  alt="Farmy McFarmson"
  className="ui centered image"
  src="farm.jpg"
/>
`;

exports[`Individual Farm Page should render the individual farm page 3`] = `
Array [
  <h2>
    Farmy McFarmson
  </h2>,
  <h2>
    Shop Coming Soon
  </h2>,
]
`;
